### ISSUE TITLE
[RFC] Docker based workflow

### ISSUE THREAD

**Issue by cserteGT3** [tlienart/Franklin.jl#455]
## Concept

Initially I was only planning submit a guide that shows, how to use a docker image to build a site and deploy on Gitlab (mentioned in https://github.com/tlienart/Franklin.jl/issues/362#issuecomment-589327038). But now that I've seen [SimpleContainerGenerator.jl](https://github.com/bcbi/SimpleContainerGenerator.jl), (that is capable of composing docker images with precompiled Julia packages), I'm thinking on a full docker based workflow, which also covers the case if one wants to build a site as part of a CI process.

I have the following in my head, assuming that the user wants to use the easiest and free resources available (eg. free GitHub/GitLab, Docker Hub):
1. One should create a docker image for their site, with SimpleContainerGenerator.jl, then push it to Docker Hub.
2. Locally one must install Docker, then it can be used to build and serve the site.
3. One commits and push changes, then CI does its job.
4. Everyone is happy (or at least me).

## Pros
- One will have a stable environment.
- One has to track only the source, not the site itself.
 - If one doesn't need a preview for a small change, just edit and commit
- One can use existing images.
- Docker: everyone ❤️  Docker.
  - Maybe julia doesn't even need to be installed, if a good image is available.

## Cons
- Docker: one must understand what it docker is, and need to install it, then register/login to Docker Hub (though I think that to use it, a beginner level is enough).
- The page must have a "stable" set of packages, so you don't have to update your image every time.
- I'm not sure about minification and pre-rendering (NodeJS.jl could work)
- Running on CI should be "fast", as GitLab and Github Actions provide 2000 mins/month for free accounts. (I think it could be a problem only if you have a lot of packages with CI running.)

## Questions
- do we need any extra code from Franklin side?

## Todos (which I'm happy to undertake)
- example scripts
- Guides and example sites for all platforms
  - GitLab
  - GitHub & Travis
  - GitHub & GitHub actions
  - any other?
- Franklin "wrapper function" if needed

I hope you like my proposal, if not, I'm going to do it for myself anyway 😉
Feel free to criticise and comment my idea, I'm not an expert with any of these tools, but I like playing with Julia and with stuff that makes Julia better.

**Comment by tlienart** [comment]
Hey @cserteGT3 , awesome, thanks for looking into this!

I'm a bit unclear on what the advantages of doing this would be. For instance on GitLab you can already call a Docker image with Julia in it, do some stuff with it (like call Franklin, deploy to `public/`) and it "just works". 

It would definitely be cool to do what you're suggesting, I'm just wondering whether how it compares with "just doing the basic stuff".

Maybe the best next step here would be that you try to run this full workflow as you have it in your head on GitLab and then that we compare the steps required to get this to work. Specifically on Gitlab, the steps are super easy so I'd be surprised one could do easier.

On GitHub it's not quite true, there's a bunch of hoops to jump in order to get the right keys at the right place so that actions can be triggered etc. But I don't see how the docker workflow would facilitate this?

Basically I think the idea may be clear in your head but not yet in mine 😂  so maybe the best is that you do a POC and chat 😄 

thanks!!

**Comment by cserteGT3** [comment]
> I'm a bit unclear on what the advantages of doing this would be. 

Me neither (yet). 😄 


> Basically I think the idea may be clear in your head but not yet in mine 😂 so maybe the best is that you do a POC and chat 😄

Agree! Will come back if I come up with something.

**Comment by terasakisatoshi** [comment]
@cserteGT3 FYI

- [my blog](https://github.com/terasakisatoshi/MathSeminar.jl) provides an example of workflow that uses Docker container. By running `make && make web`, it will generate `__site` locally.
- If one would like to deploy web site, just push commit to Github as as https://github.com/tlienart2/tlienart2.github.io suggests. Then, Github Actions runs/create `__site` automatically. It's easy.

**Comment by terasakisatoshi** [comment]
- I've confirmed I could deploy my web page using [GitHub Actions deploy.yml](https://github.com/terasakisatoshi/MathSeminar.jl/blob/master/.github/workflows/deploy.yml) that runs `docker-compose run ...`. My `docker-compose.yml` and `Dockerfile` can be found at 
  - https://github.com/terasakisatoshi/MathSeminar.jl/blob/master/docker-compose.yml
  - https://github.com/terasakisatoshi/MathSeminar.jl/blob/master/Dockerfile

- Here is my workflow, and I believe it is what @cserteGT3 want.

```console
$ git clone https://github.com/terasakisatoshi/MathSeminar.jl.git # only once
$ cd MathSeminar.jl && make # build Docker image to build my website
$ (open text editor and edit somefile.md)
$ make franklin # almost equivalent to using Franklin; serve() inside of docker container
$ git add somefile.md
$ git push origin master # GitHub Actions build website automatically
```
